File: mattress_app/public/js/bulk_create_mattress.js
/* Module 4 : "Create Mattress Model" dialog.
 *
 * Adds a button on the Item list view that opens a dialog to bulk-create a new
 * mattress model across several thicknesses with brand-first naming + area
 * pricing. Wire via hooks.py:
 *     doctype_list_js = {"Item": "public/js/bulk_create_mattress.js"}
 */

frappe.listview_settings = frappe.listview_settings || {};
frappe.listview_settings["Item"] = frappe.listview_settings["Item"] || {};

const _orig_item_onload = frappe.listview_settings["Item"].onload;

frappe.listview_settings["Item"].onload = function (listview) {
	if (_orig_item_onload) {
		try {
			_orig_item_onload(listview);
		} catch (e) {
			// don't let an existing handler block our button
			console.error(e);
		}
	}

	listview.page.add_inner_button(__("Create Mattress Model"), function () {
		open_create_model_dialog();
	});
};

function open_create_model_dialog() {
	const d = new frappe.ui.Dialog({
		title: __("Create Mattress Model"),
		size: "large",
		fields: [
			{
				fieldname: "brand",
				fieldtype: "Link",
				label: __("Brand"),
				options: "Brand",
				reqd: 1,
				description: __(
					"Pick an existing brand, or type a new name and choose 'Create a new Brand'."
				),
			},
			{
				fieldname: "model",
				fieldtype: "Data",
				label: __("Model Name"),
				description: __("e.g. Naturalite. Item code becomes 'Brand Model-<thk>MM'."),
				reqd: 1,
			},
			{
				fieldname: "item_group",
				fieldtype: "Link",
				label: __("Item Group"),
				options: "Item Group",
				default: "Products",
			},
			{
				fieldname: "stock_uom",
				fieldtype: "Link",
				label: __("UOM"),
				options: "UOM",
				default: "Nos",
			},
			{
				fieldname: "gst_hsn_code",
				fieldtype: "Data",
				label: __("HSN Code (optional)"),
			},
			{ fieldtype: "Section Break", label: __("Thicknesses & 72×72 Prices") },
			{
				fieldname: "rows",
				fieldtype: "Table",
				label: __("Thickness Rows"),
				cannot_add_rows: false,
				in_place_edit: true,
				reqd: 1,
				data: [],
				fields: [
					{
						fieldname: "thickness",
						fieldtype: "Int",
						label: __("Thickness (mm)"),
						in_list_view: 1,
						reqd: 1,
					},
					{
						fieldname: "price_72",
						fieldtype: "Currency",
						label: __("72×72 Price"),
						in_list_view: 1,
						reqd: 1,
					},
				],
			},
			{ fieldtype: "HTML", fieldname: "preview" },
		],
		primary_action_label: __("Create Items"),
		primary_action(values) {
			const rows = (values.rows || []).filter((r) => r.thickness && r.price_72);
			if (!rows.length) {
				frappe.msgprint(__("Add at least one thickness row with a price."));
				return;
			}

			frappe.call({
				method: "mattress_app.api.bulk_create_mattress.create_model",
				args: {
					brand: values.brand,
					model: values.model,
					rows: JSON.stringify(rows),
					item_group: values.item_group,
					stock_uom: values.stock_uom,
					gst_hsn_code: values.gst_hsn_code,
				},
				freeze: true,
				freeze_message: __("Creating items…"),
				callback(r) {
					if (r.message) {
						const m = r.message;
						let html = `<b>${frappe.utils.escape_html(m.family)}</b><br>`;
						if (m.created && m.created.length) {
							html +=
								__("Created:") +
								"<br>" +
								m.created
									.map((c) => "• " + frappe.utils.escape_html(c))
									.join("<br>");
						}
						if (m.skipped && m.skipped.length) {
							html +=
								"<br><br>" +
								__("Already existed (price updated):") +
								"<br>" +
								m.skipped
									.map((c) => "• " + frappe.utils.escape_html(c))
									.join("<br>");
						}
						frappe.msgprint({ title: __("Done"), message: html, indicator: "green" });
						d.hide();
						cur_list && cur_list.refresh && cur_list.refresh();
					}
				},
			});
		},
	});

	// live preview of resulting item codes
	const update_preview = () => {
		const brand = d.get_value("brand");
		const model = d.get_value("model");
		const rows = d.get_value("rows") || [];
		if (!brand || !model) {
			d.fields_dict.preview.$wrapper.html("");
			return;
		}
		const family = `${brand} ${model}`.trim();
		const lines = rows
			.filter((r) => r.thickness)
			.map(
				(r) =>
					`• <code>${frappe.utils.escape_html(family)}-${r.thickness}MM</code>` +
					(r.price_72 ? ` &nbsp;→ rate/sqin ≈ ${(r.price_72 / 5184).toFixed(4)}` : "")
			);
		d.fields_dict.preview.$wrapper.html(
			`<div class="text-muted" style="margin-top:8px;">` +
				__("Will create:") +
				`<br>${lines.join("<br>") || "—"}</div>`
		);
	};

	d.fields_dict.brand.df.onchange = update_preview;
	d.fields_dict.model.df.onchange = update_preview;
	d.fields_dict.rows.grid.wrapper.on("change", update_preview);

	d.show();
}

File: mattress_app/api/bulk_create_mattress.py
"""
Module 4 : Bulk-create a new mattress model's items across several thicknesses,
           with brand-first naming and area pricing.

Called by the "Create Mattress Model" dialog (client side). For each thickness
row it creates:
  - Thickness record (if missing)
  - Item  : code = "{Brand} {Model}-{thk}MM",  name = "{Brand} {Model}"
            item_group = "Products", brand = <brand>, custom_use_area_pricing = 1
  - Item Price (Standard Selling) = the 72x72 price entered for that thickness

Naming:
  family    = f"{brand} {model}"            e.g. "Dr. Back Naturalite"
  item_code = f"{family}-{thk}MM"           e.g. "Dr. Back Naturalite-250MM"
  item_name = family                        e.g. "Dr. Back Naturalite"

Server entry point (whitelisted), called from JS:
  mattress_app.api.bulk_create_mattress.create_model
"""

import json

import frappe

PRICE_LIST = "Standard Selling"
DEFAULT_GROUP = "Products"
DEFAULT_UOM = "Nos"


def _ensure_item_name(family):
	"""The custom_name dropdown on Quotation Item is a Link to 'Item Name'.
	Standalone items don't get this auto-created (the Item validate hook only
	does it for variants), so we create it explicitly."""
	if family and not frappe.db.exists("Item Name", {"item_name": family}):
		frappe.get_doc({"doctype": "Item Name", "item_name": family}).insert(ignore_permissions=True)


def _ensure_brand(brand):
	if brand and not frappe.db.exists("Brand", brand):
		frappe.get_doc({"doctype": "Brand", "brand": brand}).insert(ignore_permissions=True)


def _ensure_thickness(thk_value):
	"""Create a Thickness record if missing. Returns its name."""
	# Thickness records are keyed by their own naming; match on the 'value' field.
	existing = frappe.db.get_value("Thickness", {"value": str(thk_value)}, "name")
	if existing:
		return existing
	doc = frappe.get_doc({"doctype": "Thickness", "value": str(thk_value)})
	doc.insert(ignore_permissions=True)
	return doc.name


def _set_item_price(item_code, price):
	existing = frappe.get_all(
		"Item Price",
		filters={"item_code": item_code, "price_list": PRICE_LIST, "selling": 1},
		fields=["name"],
	)
	if existing:
		frappe.db.set_value("Item Price", existing[0].name, "price_list_rate", price)
	else:
		frappe.get_doc(
			{
				"doctype": "Item Price",
				"item_code": item_code,
				"price_list": PRICE_LIST,
				"selling": 1,
				"price_list_rate": price,
			}
		).insert(ignore_permissions=True)


@frappe.whitelist()
def create_model(brand, model, rows, item_group=None, stock_uom=None, gst_hsn_code=None):
	"""
	brand  : brand name (also the family prefix)
	model  : model name (e.g. 'Naturalite')
	rows   : JSON list of {"thickness": 250, "price_72": 80142}
	"""
	if isinstance(rows, str):
		rows = json.loads(rows)

	brand = (brand or "").strip()
	model = (model or "").strip()
	if not brand or not model:
		frappe.throw("Brand and Model are required.")
	if not rows:
		frappe.throw("Add at least one thickness row.")

	family = f"{brand} {model}".strip()
	item_group = item_group or DEFAULT_GROUP
	stock_uom = stock_uom or DEFAULT_UOM

	_ensure_brand(brand)
	_ensure_item_name(family)

	created = []
	skipped = []
	for r in rows:
		thk = str(r.get("thickness") or "").strip()
		price = r.get("price_72")
		if not thk or not price:
			continue

		thickness_name = _ensure_thickness(thk)

		item_code = f"{family}-{thk}MM"
		if frappe.db.exists("Item", item_code):
			skipped.append(item_code)
			# backfill the thickness link + refresh price
			frappe.db.set_value(
				"Item", item_code, "custom_thickness_link", thickness_name, update_modified=False
			)
			_set_item_price(item_code, price)
			continue

		doc = frappe.new_doc("Item")
		doc.item_code = item_code
		doc.item_name = family  # family only (thickness lives in code)
		doc.item_group = item_group
		doc.stock_uom = stock_uom
		doc.brand = brand
		doc.custom_use_area_pricing = 1
		doc.custom_thickness_link = thickness_name
		doc.has_variants = 0
		doc.variant_of = None
		if gst_hsn_code:
			doc.gst_hsn_code = gst_hsn_code
		doc.insert(ignore_permissions=True)

		_set_item_price(item_code, price)
		created.append(item_code)

	frappe.db.commit()
	return {
		"family": family,
		"created": created,
		"skipped": skipped,
		"message": f"Created {len(created)} item(s)"
		+ (f", updated price on {len(skipped)} existing" if skipped else ""),
	}
